doc/decisions/DR023-Versioning-traits-and-specifications.md

```diff
@@ -0,0 +1,702 @@
+thing would be to enhance the low level `TraitsData` type to have a
+first class understanding of versioning, this cannot be so because not
+every method takes a `TraitsData` as an argument. The lowest-friction
+course of action might be to pack the version into `TraitId`, essentially
+doing string-munging to communicate the version across the api-boundary.
+
+Hosts and managers would need to be inspect the version data, which we
+could provide conveniences for, and construct the correct version
+of the trait view.
+
+```yml
+# traits.yml
+Camera:
+  description: >
+    The entity represents a device that captures a view through
+    re-projection or encoding into some other form.
+  usage:
+    - entity
+  version: 3
+```
+
+```python
+# Manager
+# In this example, CameraTrait is on version 3
+# SpatialTrait is on version 2
+#
+# The manager is responding to a resolve, needing to populate a
+# TraitsData with a camera trait and a spatial trait of the expected
+# version
+cameraVersion = mediacreation.versionFromTraitId(mediacreation.CameraTrait.kId)
+spatialVersion = mediacreation.versionFromTraitId(mediacreation.SpatialTrait.kId)
+
+# Support all known major versions.
+# Camera
+success_result = TraitsData()
+if (cameraVersion >= 3 and cameraVersion < 4):
+     trait = mediacreation.CameraTrait(success_result)
+     ... # Up to date camera population
+elif (cameraVersion >= 2 and cameraVersion < 1):
+     trait = mediacreation.legacy.CameraTrait_v2(success_result)
+     ... # V2 camera population
+elif (cameraVersion >= 1):
+     trait = mediacreation.legacy.CameraTrait_v1(success_result)
+     ... # V1 camera population
+else:
+  raise RuntimeError()
+
+# Spatial
+if (spatialVersion >= 2 and spatialVersion < 3):
+    trait = mediacreation.SpatialTrait(success_result)
+    ... # Up to date spatial population
+elif (spatialVersion >= 1):
+    trait = mediacreation.legacy.SpatialTrait_v1(success_result)
+     ... # V1 to date spatial population
+else:
+  raise RuntimeError()
+```
+
+### Benefits
+- Maximally flexible, hosts and managers can disregard versioning for
+  traits they don't care about.
+- Obvious from the YAML which version a trait is.
+- Individual traits will only update version when they change, thus they
+  will not have "gaps", they will always go v1->v2->v3.
+- A single trait can be updated in isolation without any effect on other
+  traits in the package.
+- Semver could be used to provide additional runtime compatibility
+  information.
+
+### Drawbacks
+- Version has to be encoded in `traitID` because several methods
+  (including `resolve`) take `traitSet`, which is only a collection of
+  `traitIDs`. There is no `TraitsData` to inspect. This would be a major
+  breaking change as currently conceived.
+- Version support may become fragmented if hosts and managers provide
+  versioned support for only certain traits
+- When packing multiple traits, manager has to branch on each trait as
```

Comment:
If we used SemVer or similar, and had the concept of none breaking changes, it feels to me like branching could be minimal for most traits.

Reply:
In OTIO they don't even change the version for non-breaking changes.

I think the problem here is that you can't have one big outer branch on trait package version, and instead have to check the version individually for every trait.

Reply:
Yeah you'd likely reuse code for traits that haven't changed, but how you do that is more constrained with per-trait versioning.  

E.g. with per-trait versions you must branch around every trait when you receive it, leading to fragmented handler code, whereas with a global schema version you could maintain a single handler (using OO techniques to reuse code) dedicated to a specific schema version.

Reply:
I guess I was thinking that you it probably be unlikely to have a single handler per-package.